### PR TITLE
pilot: fix bug in endpoint updates

### PR DESCRIPTION
**Pull request by howardjohn**
TBH I am not sure how this isn't a bigger issue. Currently if an
endpoint changes we are not triggering a push at all. In this case, it
was port number, but could be others.

Implementation is *bad* right now, but if this is desired we can swap
for an optimized Equals implementation

### CODE REVIEW

**Comment by howardjohn** [comment]
@hzxuzhonghu @ramaraochavali wdyt? if desired ill write an optimized Equals function

File: pilot/pkg/serviceregistry/serviceregistry_test.go [hunk 1]
```diff
@@ -241,10 +242,13 @@ func TestWorkloadInstances(t *testing.T) {
...
+		// Endpoint update is triggered since its a brand new service
+		if ev := fx.WaitOrFail(t, "xds full"); !ev.Reason.Has(model.EndpointUpdate) {
+			t.Fatalf("xds push reason does not contain %v: %v", model.EndpointUpdate, ev)
+		}
 		// headless service update must trigger nds push, so we trigger a full push.
-		ev := fx.WaitOrFail(t, "xds full")
-		if !ev.Reason.Has(model.HeadlessEndpointUpdate) {
-			t.Fatalf("xds push reason does not contain %v", model.HeadlessEndpointUpdate)
+		if ev := fx.WaitOrFail(t, "xds full"); !ev.Reason.Has(model.HeadlessEndpointUpdate) {
+			t.Fatalf("xds push reason does not contain %v: %v", model.HeadlessEndpointUpdate, ev)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/serviceregistry_test.go:251]: This is NOT a production impact change. Only test.

Now our test properly mimics how the real istiod does pushes; in the real world we **are**  triggering double pushes.

Note with debounce its mostly moot, though

**Review by ramaraochavali — APPROVED** [review]

**Pull request merged**